Unify OpenSimScene with OpenSimGUIScene

AlbertoCasasOrtiz opened the pull request:
Substitutes OpenSimScene by OpenSimGUIScene when using the browser. The GUI should not have changed.

Created SimpleModelScene to show previsualizations of models in the `/models` page more efficiently.

In the browser, the URI of the latest loaded model is stored in local storage, so there is persistence, and latest visualized model is loaded automatically.

Additional fixes:
 - Fixes spacing issues when using the tree between GUI and browser modes.
 - Fixes https://github.com/opensim-org/opensim-gui/issues/1629 (See changes in `src/components/pages/ModelViewPage.tsx`)
 - Fixes #355 (See this commit: https://github.com/opensim-org/opensim-viewer/pull/304/changes/7577e294b2b25c0e9d76e4696474a330e8329a5c)
 - Fixes https://github.com/opensim-org/opensim-gui/issues/1619 (See this commit: https://github.com/opensim-org/opensim-viewer/pull/304/commits/fde096afbe9ac9591e63a33a48a23c4b616666f6)
 
 
 For testing:
 
 1. Check that the visualizer works properly in the browser, allowing to visualize prebuilt models.
 1. Check that the loading component (when loading the viewer) appears properly centered in both browser and GUI.
 2. Check that if no animation available, it will show a toast notification indicating there is no animation (instead of silently failing in the GUI and showing uncaught exception in browser).
 3. Check that when recording any video, it always shows the opensim watermark at bottom left.
 4. Pass through our testing sheet in the GUI, to check that we have not affected other behaviour or introduced any other issue.
 
 Known issues:
 
- Looks like gltf models cannot be loaded in the browser through the DropFile class. This only affect the browser, as we do not use it in the GUI. This may have been there for a while --> Should I create a new issue for this and address separately?

AlbertoCasasOrtiz committed:
Unifying OpenSimScene and OpenSimGUIScene.

AlbertoCasasOrtiz committed:
Merge branch 'dev' into unify-opensimscene

Signed-off-by: Alberto <albertocasasortiz@gmail.com>

AlbertoCasasOrtiz committed:
Unified OpenSimGUIScene with OpenSimGUI. Added Preview Scene. Fixed loading model issues.

AlbertoCasasOrtiz commented:
@aymanhab This should be ready now. Can you test if this works for you in the GUI?

aymanhab commented:
@AlbertoCasasOrtiz Sure, will review once we get the GUI beta build going so we don't add extra change

AlbertoCasasOrtiz committed:
Merge pull request #331 from opensim-org/dev

Update unify-opensimscene branch

aymanhab commented:
@AlbertoCasasOrtiz I know this has been painfully done multiple times, but is there a version of this working with the dev branch used in 4.6 beta (aka opensim_46 branch) to revive the online viewer without conflicts? Thank you

AlbertoCasasOrtiz commented:
I am not sure, I will try to fix merge conflicts and test, so we can close this.

AlbertoCasasOrtiz committed:
Renamed OpenSimScene to SimpleModelScene

AlbertoCasasOrtiz committed:
Unified GUI and viewer. Fixed spacing issues with tree and loader.

AlbertoCasasOrtiz committed:
Capturing error when no animation.

AlbertoCasasOrtiz committed:
Merge pull request #356 from opensim-org/unify-viewer-gui

Unify viewer gui

AlbertoCasasOrtiz commented:
Hi @aymanhab. This should be ready for testing. It has the changes related to unifying browser and viewer, and small additional fixes. For additional fixes, I have pointed the file and/or commit that changed.

AlbertoCasasOrtiz committed:
Added watermark to videos.

AlbertoCasasOrtiz committed:
Merge pull request #357 from opensim-org/unify-viewer-gui

Added watermark to videos.

aymanhab commented:
Thanks @AlbertoCasasOrtiz will build locally and test. Some questions:
1. Did you build the GUI with it and all the functionality works?
2. Most of the changes are to classes not used in the GUI except for OpenSimLoader which is at the core of the GUI interface, can you describe what needed to be changed there and why?
Thank you

AlbertoCasasOrtiz committed:
Fixed issue with gltf files manually uploaded through DropFile.tsx.

AlbertoCasasOrtiz commented:
@aymanhab I am actually testing right now, trying to follow the spreadsheets.

> 1. Did you build the GUI with it and all the functionality works?

Yes, I built with the GUI and what I have tested so far works, but I am still going through the spreadsheet.

> 2. Most of the changes are to classes not used in the GUI except for OpenSimLoader which is at the core of the GUI interface, can you describe what needed to be changed there and why?
Thank you

Yes, I did this long ago, so I tried to see the changes in git diff, but looks like it detected formatting changes as actual changes. After reviewing the code, the changes to this file are core to unifying the two files and make everything work fine in GUI and browser. The real changes are:

1. I added a constructor: https://github.com/opensim-org/opensim-viewer/blob/0ffbdbc1bbcbf9b28e44297caf31e75f6eeb9a72/src/state/OpenSimLoader.js#L31
so when used in `OpenSimGUIScene`, it can manage gltf files directly, in contrast to previous version of OpenSimGUIScene that did not accept gltf/glb as input (required in the viewer if we load locally a file as we did before). https://github.com/opensim-org/opensim-viewer/blob/0ffbdbc1bbcbf9b28e44297caf31e75f6eeb9a72/src/components/Components/OpenSimGUIScene.tsx#L42
I also added some helper methods for this behaviour: https://github.com/opensim-org/opensim-viewer/blob/0ffbdbc1bbcbf9b28e44297caf31e75f6eeb9a72/src/state/OpenSimLoader.js#L109
And now it directly checks if the file is gltf and loads it. If not, it will procceed as normal as in the GUI:
https://github.com/opensim-org/opensim-viewer/blob/0ffbdbc1bbcbf9b28e44297caf31e75f6eeb9a72/src/state/OpenSimLoader.js#L43
If I am not wrong, all of the extra behaviour was previously specific to OpenSimScene, and this is the reason OpenSimScene and OpenSimGUIScene were incompatibles. While OpenSinGUIScene only supported scenes as JSON objects, this version of the loader also supports gltf.
Thanks to this question I was able to find what was happening with gltf files manually uploaded through browser and fix it.
2. Looks like I also changed the layering here from 2 to 3: https://github.com/opensim-org/opensim-viewer/blob/0ffbdbc1bbcbf9b28e44297caf31e75f6eeb9a72/src/state/OpenSimLoader.js#L561
I don't remember exactly why, but must be related to elements shown in the videos. This specific change may not be necessary anymore.

AlbertoCasasOrtiz committed:
Merge pull request #358 from opensim-org/unify-viewer-gui

Fixed issue with gltf files manually uploaded through DropFile.tsx.

AlbertoCasasOrtiz committed:
Increased margings of watermark in videos.

AlbertoCasasOrtiz committed:
Merge pull request #359 from opensim-org/unify-viewer-gui

Increased margings of watermark in videos.

aymanhab commented:
I'd reverse the layers change unless really needed.

aymanhab commented:
Also is this  mt.json the same as the one used in the GUI, can we unify/reuse? If not rename so distinction is clear

AlbertoCasasOrtiz committed:
Reversed layer change.

AlbertoCasasOrtiz committed:
Deleted extra mt.json file.

AlbertoCasasOrtiz commented:
@aymanhab I reversed the layer change.

Regarding the `mt.json` file, the original one is in a folder called `/builtin`, and there is another inside of `/public`. The extra one I added inside of `/public/builtin/` for testing I just removed it.

I was wondering what the one  inside of `/builtin` is used for as both are the same file with same uuid.

AlbertoCasasOrtiz commented:
@aymanhab Testing on my side for the GUI looks good. Let me know if you have any other comment on this PR, or feel free to merge when ready.

aymanhab commented:
Thanks @AlbertoCasasOrtiz yes the GUI looks fine, if you tested the standalone viewer let me know what works and what doesn't and I'll merge then we can refine the standalone viewer as we go. Thank you

AlbertoCasasOrtiz commented:
Sure, these are the things I've tested on the browser that are not working as expected. I was going to create separate issues for it once this is closed:

1. We have a button on top bar to open the viewer. If you open a model, change to another page (e.g. Model Galledy), and then go back to viewer, you will not see the previously loaded model. I saved it's path in local browser storage, and there is persistence now. However, there are two errors associated with this:
    a. If the model was loaded locally, it will show a "Could not load blob: Failed to fetch" error, since the model is not persistent when uploaded as blob in the browser.
    b. We have no current mechanism for closing a model as we do in the GUI, so if you load a persistent model (e.g., one of those in the Model Gallery), the only current way of closing it is by opening a different standalone model from model gallery.
    c. You can load multiple models from gltf files in the same viewer instance, however, if you do it with a model from Model Gallery, since the gltf model loaded manually was not persistent, it will fail when coming back to the viewer.
    
I think these are easy to solve, by providing persistence in some way (we did it in the past by uploading the model to AWS, although we can also save only the latest model in local storage), and allowing to close a model from the Tree as we do in the GUI.

aymanhab commented:
Thanks @AlbertoCasasOrtiz I'll go ahead and merge so we have a unified environment, and please open issues for the above or one place holder that refers to these so they don't fall through the cracks. 🥇

AlbertoCasasOrtiz commented:
Done! I've created two new issues
